internal/devices/assigned_devices.go

```diff
@@ -0,0 +1,163 @@
+// +build windows
+
+package devices
+
+import (
+	"context"
+	"fmt"
+	"io/ioutil"
+	"net"
+	"strings"
+
+	winio "github.com/Microsoft/go-winio"
+	"github.com/Microsoft/go-winio/pkg/guid"
+	"github.com/Microsoft/hcsshim/internal/cmd"
+	"github.com/Microsoft/hcsshim/internal/log"
+	"github.com/Microsoft/hcsshim/internal/resources"
+	"github.com/Microsoft/hcsshim/internal/shimdiag"
+	"github.com/Microsoft/hcsshim/internal/uvm"
+	specs "github.com/opencontainers/runtime-spec/specs-go"
+	"github.com/pkg/errors"
+)
+
+// HandleAssignedDevicesWindows does all of the work to setup the hosting UVM and retrieve
+// device information for adding assigned devices on a WCOW container definition.
+//
+// First, devices are assigned into the hosting UVM. Drivers are then added into the UVM and
+// installed on the matching devices. This ordering allows us to guarantee that driver
+// installation on a device in the UVM is completed before we attempt to create a container.
+//
+// Then we find the location paths of the target devices in the UVM and return the results
+// as WindowsDevices.
+func HandleAssignedDevicesWindows(ctx context.Context, vm *uvm.UtilityVM, annotations map[string]string, specDevs []specs.WindowsDevice) (resultDevs []specs.WindowsDevice, closers []resources.ResourceCloser, err error) {
```

Comment:
I think the pieces of code that specifically deal with the OCI spec should be in `hcsoci`. The way I would think about the separation is like this:
- `devices` package: Assuming you've never heard of OCI and have no idea what's in the spec, figure out what you want the interface to look like for adding devices to a UVM/container.
- `hcsoci` package: Take what we have in the OCI spec, and map it to the concepts exposed by `devices`.

As a strawman example:
- `devices` could expose a `AddDevice(ctx context.Context, uvm *uvm.UtilityVM, pciID string, drivers []string, deviceUtilTool string)`.
- `hcsoci` could examine the set of `WindowsDevice` entries in the spec, along with the annotation instructing what drivers to install, and use that information to construct a call to `AddDevice`.

Reply:
My hope is we can remove the `specs "github.com/opencontainers/runtime-spec/specs-go"` import from this package. :)

Reply:
Updated :) Let me know how you feel about the new change.